src/lib/CoreInstance.cc

```diff
@@ -0,0 +1,304 @@
+
+  return;
+}
+
+void CoreInstance::createProcessMemory() {
+  // Get the process image and its size
+  processMemory_ = process_->getProcessImage();
+  processMemorySize_ = process_->getProcessImageSize();
+}
+
+void CoreInstance::createL1InstructionMemory(
+    const simeng::MemInterfaceType type) {
+  // Create a L1I cache instance based on type supplied
+  if (type == simeng::MemInterfaceType::Flat) {
+    instructionMemory_ = std::make_shared<simeng::FlatMemoryInterface>(
+        processMemory_.get(), processMemorySize_);
+  } else if (type == simeng::MemInterfaceType::Fixed) {
+    instructionMemory_ = std::make_shared<simeng::FixedLatencyMemoryInterface>(
+        processMemory_.get(), processMemorySize_,
+        config_["LSQ-L1-Interface"]["Access-Latency"].as<uint16_t>());
+  } else {
+    std::cerr << "Unsupported memory interface type used in "
+                 "createL1InstructionMemory()."
+              << std::endl;
+    exit(1);
+  }
+
+  return;
+}
+
+void CoreInstance::setL1InstructionMemory(
+    std::shared_ptr<simeng::MemoryInterface> memRef) {
+  // Set the L1I cache instance to use
+  instructionMemory_ = memRef;
+  return;
+}
+
+void CoreInstance::createL1DataMemory(const simeng::MemInterfaceType type) {
+  // Create a L1D cache instance based on type supplied
+  if (type == simeng::MemInterfaceType::Flat) {
+    dataMemory_ = std::make_shared<simeng::FlatMemoryInterface>(
+        processMemory_.get(), processMemorySize_);
+  } else if (type == simeng::MemInterfaceType::Fixed) {
+    dataMemory_ = std::make_shared<simeng::FixedLatencyMemoryInterface>(
+        processMemory_.get(), processMemorySize_,
+        config_["LSQ-L1-Interface"]["Access-Latency"].as<uint16_t>());
+  } else {
+    std::cerr
+        << "Unsupported memory interface type used in createL1DataMemory()."
+        << std::endl;
+    exit(1);
+  }
+
+  return;
+}
+
+void CoreInstance::setL1DataMemory(
+    std::shared_ptr<simeng::MemoryInterface> memRef) {
+  // Set the L1D cache instance to use
+  dataMemory_ = memRef;
+  return;
+}
+
+void CoreInstance::createCore() {
+  // If memory interfaces must be manually set, ensure they have been
+  if (setDataMemory_ && dataMemory_ == nullptr) {
+    std::cerr << "Data memory not set. External Data memory must be manually "
+                 "set using the setL1DataMemory(...) function."
+              << std::endl;
+    exit(1);
+  } else if (setInstructionMemory_ && instructionMemory_ == nullptr) {
+    std::cerr << "Instruction memory not set. External instruction memory "
+                 "interface must be manually set using the "
+                 "setL1InstructionMemory(...) function."
+              << std::endl;
+    exit(1);
+  }
+
+  // Construct architecture object
+  arch_ =
+      std::make_unique<simeng::arch::aarch64::Architecture>(kernel_, config_);
+
+  // Construct branch predictor object
+  predictor_ = std::make_unique<simeng::GenericPredictor>(config_);
```

Comment:
Given there are multiple branch predictor classes available, could a config option be introduced to define which one is used?

Reply:
We could. I think such functionality would be better suited to an upcoming PR that targets Branch Predication/speculation in general.